server/src/main/java/io/crate/planner/optimizer/rule/MoveLimitBeneathUnion.java

```diff
@@ -64,27 +64,32 @@ public LogicalPlan apply(Limit limit,
                              NodeContext nodeCtx,
                              Function<LogicalPlan, LogicalPlan> resolvePlan) {
         Union union = captures.get(unionCapture);
+        var newLimit = getGlobalLimit(limit);
+        var newOffSet = Literal.of(0);
+
+        return new Limit(
+            new Union(
+                new Limit(union.lhs(), newLimit, newOffSet, false, false),
+                new Limit(union.rhs(), newLimit,newOffSet, false, false),
+                union.outputs()
+            ),
+            limit.limit(),
+            limit.offset(),
+            true,
+            limit.isPushedBeneathJoin()
+        );
+    }
+
+    static Literal<Integer> getGlobalLimit(Limit limit) {
         // We cannot push the offset down to the sources, otherwise
-        // we may loose rows, therefore we create a new limit where
+        // we may lose rows, therefore we create a new limit where
         // the offset is added to the limit.
         if (limit.limit() instanceof Literal<?> l &&
```

Comment:
can we use `Symbol#isLiteral` maybe a new method without datatype arg, to avoid `instanceof`?